Data/Aeson/TH.hs

```diff
@@ -349,11 +354,35 @@ consToValue target jc opts instTys cons = do
               , let conName = constructorName con
               ]
         | otherwise -> [argsToValue target jc tvMap opts True con | con <- cons]
+    lets :: ToJSONFun -> ExpQ -> ExpQ
+    lets Value e = e
+    lets (Encoding r) e = do
+      e' <- e -- collect used keys
+      labels <- runIO $ readIORef r
+      let decs = map
+            (\k -> let str = BL.unpack $ B.toLazyByteString $ EB.string k
+                  in valD (varP $ varOfKey k) (normalB [| S.pack str |]) [])
+            (Set.toList labels)
+      letE decs $ pure e'
+
+varOfKey :: String -> Name
+varOfKey l = mkName $ "bs_" ++ l
+
+type KeySet = IORef (Set String)
```

Comment:
`IORef`? Why?

Reply:
I use the `IORef` to store the JSON strings encountered during the generation of `toEncoding`. In the end, I gather the strings from the `IORef` and create the top-level bindings.

What doesn't look good about the `IORef`? If we don't want it, I guess the alternative is using a state monad style, but the refactoring seems quite invasive.

Reply:
I feel uneasy.

There is a non-local state, and it's hidden. I don't know if changing `type KeyMap` to `type KeyMapIORef` and adding a note about the scheme will help.

(I'm not opposed to having `IORef` based `StateT s IO`-like monads, but then I'd like to have that monad in `newtype`, and the problem of refactoring is still there).

What others think, @bergmark @RyanGlScott

Reply:
Do you really save much compile time with that optimization? The strings involved here are short and used only once, aren't they?

Reply:
@Lysxia as you can see from benchmark in PR description, GHC is dumb, and doesn't float bytestring `Encoding`s. So this is welcome change.

Using `IORef` is *easy* way to introduce state into TH structure traversal, so we can do manual float-out. It's not a performance optimization, but a "as Q is IO, let's embrace it and make our lives easier but changing everything".
I understand that viewpoint. However I'm not sure if it makes our lives easier in long term.

Reply:
Thanks for the clarification. I had indeed misunderstood the change.

I also found `IORef` a bit surprising but ultimately it doesn't seem worse than the alternative of `StateT KeyMap` (without `IORef`). The state is arguably quite "well-behaved" since it's only accessible when `r :: KeyMap` is in scope and all updates are done by `keyRef` where it's easy to see that no overwriting happens. It's completely contained in this file and invisible to users, so I think this situation is really far from the horror tales of global state programmers like to tell.

Reply:
I renamed `KeyMap` to `KeyMapIORef`.